shared/modules/transaction.utils.js

```diff
@@ -131,3 +132,79 @@ export async function determineTransactionType(txParams, query) {
 
   return { type: result, getCodeResponse: contractCode };
 }
+
+const INFERRABLE_TRANSACTION_TYPES = [
+  TRANSACTION_TYPES.TOKEN_METHOD_APPROVE,
+  TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER,
+  TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER_FROM,
+  TRANSACTION_TYPES.CONTRACT_INTERACTION,
+  TRANSACTION_TYPES.SIMPLE_SEND,
+];
+
+/**
+ * Given a transaction meta object, determine the asset type that the
+ * transaction is dealing with, as well as the standard for the token if it
+ * is a token transaction.
+ *
+ * @param {import('../constants/transaction').TransactionMeta} txMeta -
+ *  transaction meta object
+ * @param {EthQuery} query - EthQuery instance
+ * @param {Function} getTokenStandardAndDetails - function to get token
+ *  standards and details.
+ * @returns {{ assetType: string, tokenStandard: string}}
+ */
+export async function determineTransactionAssetType(
+  txMeta,
+  query,
+  getTokenStandardAndDetails,
+) {
+  // If the transaction type is already one of the inferrable types, then we do
+  // not need to re-establish the type.
+  let inferrableType = txMeta.type;
+  if (INFERRABLE_TRANSACTION_TYPES.includes(txMeta.type) === false) {
+    // Because we will deal with all types of transactions (including swaps)
+    // we want to get an inferrable type of transaction that isn't special cased
+    // that way we can narrow the number of logic gates required.
+    const result = await determineTransactionType(txMeta.txParams, query);
+    inferrableType = result.type;
+  }
+
+  const isTokenMethod = [
+    TRANSACTION_TYPES.TOKEN_METHOD_APPROVE,
+    TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER,
+    TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER_FROM,
+  ].find((methodName) => isEqualCaseInsensitive(methodName, inferrableType));
+
+  if (isTokenMethod) {
```

Comment:
There are many possible cases where we could still get the `tokenStandard` where the particular method being called is not included in `isTokenMethod`. If a user calls any standard or non-standard method on an ERC20, ERC721 or ERC1155 token contract we can derive the assetType just by passing the tokenAddress (the second arg on getTokenStandardAndDetails should be optional and I will open a PR so that the function signature reflects that). `getTokenStandardAndDetails` will throw the error `Unable to determine contract standard` if the contract is not one of these asset types and then we can set it to unknown.

Reply:
PR for the optional arg piece here: https://github.com/MetaMask/controllers/pull/717

Reply:
@adonesky1 i'm confused by this comment. Are you suggesting there are other transaction types that are in fact transactions sending or receiving tokens? I know you suggested a new item for the array that we are looking for in isTokenMethod.

Reply:
are you suggesting I could just call getTokenStandardAndDetails without first checking the transaction type?

Reply:
Ok so running back through this the solution I'm proposing is simply to add `TRANSACTION_TYPES.CONTRACT_INTERACTION` to the `isTokenMethod` array. There are many methods on the token standards that move through the transaction controller that are not `transfer` and `transferFrom`, as just a quick example `DRIP` or `DRAIN` on this [contract](https://rinkeby.etherscan.io/token/0xf5de760f2e916647fd766b4ad9e85ff943ce3a2b?a=0x806627172aF48bD5B0765D3449A7DEf80d6576ff#writeContract). There will of course be contract interactions that are not with contracts conforming to the ERC721, ERC1155, ERC20 standards and for those `getTokenStandardAndDetails` will throw an error, `details.standard` will be undefined and we will fall through to the `if (inferrableType === TRANSACTION_TYPES.CONTRACT_INTERACTION) {` block on line 203. But we might as well check for the token standard on all contract interactions and add the standard to our metrics where we can?

Reply:
Gotcha. Sounds good to me.

Reply:
@adonesky1 fixed this in the latest, let me know what you think. Added a few more inline comments too

Reply:
Looks like you've got to rebase and then 💯

Reply:
Rebased @adonesky1